### PR TITLE
RUN-2948: Refac API message map

### PR DESCRIPTION
**Pull request by joneit**
**This PR moves permissions previously defined in `actionAPINameMap` in _api_protocol_base.ts_ to the local `actionMap`.** As before, this map is filled by calls to `registerActionMap` from files in the _api_handlers_ folder. The permission defintions have therefore been moved to these files.

### Typedefs
I added some type definitions in  _api_transport_base.ts:_ `actionMap` is an `ActionMap`, made up `Endpoint`s, each of which in turn consists of an `Actor` (the API method to call) and an `ApiPath` (Steve says more are coming).

`ApiPath`, formerly an array of strings (the path elements) is not a string in dot notation syntax.

 These type names are up for discussion.

### Readability
Solely for code readability, `registerActionMap`'s input is a `ActionSpecMap`, made up of `EndpointSpec`s, which differ from `ActionMap`'s `Endpoint`s in that they can be either an `Actor` _or_ and `Endpoint`. This makes the map backwards compatible to the prior form, when there are no other properties to specify. **TL;DR the following form (which represents the majority of the definitions) still works:**
```js
'get-window-info': getWindowInfo
```

### Convenience
I added a convenience feature to `registerActionMap`:  _On input_ to `registerActionMap`, `ApiPath`s may be relative to the base given in the call. In the following example, the full `apiPath` becomes `'System.getMousePosition'`:
```js
var specs = {
        'get-mouse-position': { actor: getMousePosition, apiPath: '.getMousePosition' },
};
registerActionMap(specs, 'System');
```

**TL;DR Regardless of the input to `registerActionMap`, the output to `actionMap` is always a uniform hash of `EndPoint`s with fully qualified API paths.**

#### Beautifier options
Now accommodates beautifier options. I added one option to prevent splitting simple object literals into multiple lines. Advantages:
* Single-line object literals stay that way (as in the previous example above)
* `import` statements be one-liners again
* Makes destructuring assignments much more succinct and readable

#### License checks
White space no longer significant.

### Dueling interfaces
We have "dueling interfaces" for ack and nack payloads and functions:
* `./src/browser/api_protocol/transport_strategy/ack.ts`<br>&nbsp; &nbsp; _vs._
* `./src/shapes.ts`

We have various `Identity` interfaces:
* Where they legitimately differ they should be renamed accordingly.
* Where they are the same or could be semantically overloaded with `?` (optional member) feature, they should be merged.

### CODE REVIEW

**Comment by whyn07m3** [comment]
TL;DR (too long didn't review)

**Comment by joneit** [comment]
Rebased to `6.49.21.10`.

**Comment by joneit** [comment]
Of the [3 failed tests](https://testing-dashboard.openfin.co/#/app/sessions/api/completed/592095315114ac10ca6db447), all passed on manual retry in "quick testing" panel (except `Notification.close (from onShow)`, which I didn't try).

**Comment by joneit** [comment]
Rebased to `develop` (tip = RUN-2957).

**Comment by joneit** [comment]
Of the [2 failed tests](https://testing-dashboard.openfin.co/#/app/sessions/api/completed/59246bd15114ac10ca6db44d), one of which was _Notification.close (from onShow)_ and the other of which passed on manual retry.

File: src/browser/api_protocol/api_handlers/api_protocol_base.ts [hunk 1]
```diff
@@ -19,11 +19,13 @@ const WebSocketStrategy = require('../transport_strategy/ws_strategy').WebSocket
 const ElipcStrategy = require('../transport_strategy/elipc_strategy').ElipcStrategy;
 
 import { default as RequestHandler } from '../transport_strategy/base_handler';
-import { ActionMap, MessagePackage } from '../transport_strategy/api_transport_base';
+import { MessagePackage } from '../transport_strategy/api_transport_base';
 import { registerMiddleware as registerMeshMiddleware } from './mesh_middleware';
 import { meshEnabled } from '../../connection_manager';
+import { ApiPath, Actor, EndpointSpec, ActionSpecMap, Endpoint, ActionMap } from '../shapes';
+
+export const actionMap: ActionMap = Object.create(null); // no prototype to avoid having to call hasOwnProperty()
```
Comment (datamadic) [inline src/browser/api_protocol/api_handlers/api_protocol_base.ts]: ❓ can we use a `Map` here?

File: src/browser/api_protocol/api_handlers/api_protocol_base.ts [hunk 2]
```diff
@@ -35,12 +37,50 @@ const webSocketStrategy = new WebSocketStrategy(actionMap, requestHandler);
...
+// `actionAPINameMap` has been removed in favor of a complex `actionMap`.
+// The API paths previously defined here (as string arrays) have been incorporated into `actionMap`
+// as `ApiPath` types (strings in dot notation) in calls to `registerActionMap`.
+// To see the full list search (globally) for "apiPath:"
+
+export function registerActionMap(
+    specs: ActionSpecMap,
+    authorizationPathPrefix?: string // dot notation
+) {
+    Object.getOwnPropertyNames(specs).forEach((key: string) => {
```
Comment (datamadic) [inline src/browser/api_protocol/api_handlers/api_protocol_base.ts]: If this was a `Map` the we could just iterate over it...

File: src/browser/api_protocol/api_handlers/api_protocol_base.ts [hunk 3]
```diff
@@ -35,12 +37,50 @@ const webSocketStrategy = new WebSocketStrategy(actionMap, requestHandler);
...
+        if (!actionMap[key]) {
+            const spec: EndpointSpec = specs[key];
+            let endpoint: Endpoint;
+
+            // resolve `spec` overloads: Actor (function) vs. Endpoint (object containing `actor` function)
+            if (typeof spec === 'function') {
+                endpoint = { actor: <Actor>spec };
```
Comment (datamadic) [inline src/browser/api_protocol/api_handlers/api_protocol_base.ts]: Prefer assignment over casting as you will get better type checking

File: src/browser/api_protocol/api_handlers/api_protocol_base.ts [hunk 4]
```diff
@@ -35,12 +37,50 @@ const webSocketStrategy = new WebSocketStrategy(actionMap, requestHandler);
...
+            } else if (typeof spec === 'object') {
+                endpoint = <Endpoint>spec;
```
Comment (datamadic) [inline src/browser/api_protocol/api_handlers/api_protocol_base.ts]: Prefer assignment over casting as you will get better type checking

File: src/browser/api_protocol/api_handlers/api_protocol_base.ts [hunk 5]
```diff
@@ -35,12 +37,50 @@ const webSocketStrategy = new WebSocketStrategy(actionMap, requestHandler);
...
+            if (typeof endpoint.actor !== 'function') {
+                throw new Error(`Expected endpoint.actor to be function but found ${typeof endpoint.actor}`);
+            }
+
+            let apiPath: ApiPath = endpoint.apiPath;
+            if (apiPath) {
+                if (apiPath[0] === '.') {
+                    // leading dot appends provided authorization prefix
+                    if (!authorizationPathPrefix) {
+                        throw new Error('Needed authorization prefix not provided on this call to registerActionMap');
```
Comment (datamadic) [inline src/browser/api_protocol/api_handlers/api_protocol_base.ts]: ❓where is this caught?

File: src/browser/api_protocol/transport_strategy/ack.ts [hunk 6]
```diff
@@ -15,29 +15,32 @@ limitations under the License.
...
         this.data = data;
     }
 }
 
 export class NackPayload {
-    public readonly success: boolean = false;
-    public reason: string = '';
-    public error: Error = null;
+    public success: EverFalse;
+    public reason?: string = '';
```
Comment (datamadic) [inline src/browser/api_protocol/transport_strategy/ack.ts]: Why have these become optional?

File: src/browser/api_protocol/shapes.ts [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+ limitations under the License.
+ */
+
+// todo: we should only have one ack func and one nack func!!
+
+import { AckFunc, NackFunc } from './transport_strategy/ack';
+import { Identity, Acker, Nacker } from '../../shapes';
+
+export interface Actor {
+    (identity: Identity, data: any, ack?: Acker | AckFunc, nack?: Nacker | NackFunc): void;
```
Comment (nicholasdgoodman) [inline src/browser/api_protocol/shapes.ts]: Actor is too generic of a name, similar to "Do-er", "Thing", ...
A delegate type declaration name should indicate the type of function it represents.

Suggest: `ApiHandlerFunc`, or `ApiHandler`, or `ApiMessageHanlder`, etc...

File: src/browser/api_protocol/shapes.ts [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+}
+
+export type ApiPath = string; // path in dot notation
+
+export type Endpoint = {
+    actor: Actor;
```
Comment (nicholasdgoodman) [inline src/browser/api_protocol/shapes.ts]: Function names should be verbs.

Suggest: `HandleMessage` or similar... (see suggested type name revision, above)

File: Gruntfile.js [hunk 9]
```diff
@@ -186,13 +194,12 @@ module.exports = (grunt) => {
...
-                src: ['src/**/*.js', 'index.js']
+                src: ['src/**/*.js', 'index.js'],
+                options: beautifierOptions
             },
             'git-pre-commit': {
                 src: ['src/**/*.js', 'index.js'],
-                options: {
-                    mode: 'VERIFY_ONLY'
-                }
+                options: Object.assign({ mode: 'VERIFY_ONLY' }, beautifierOptions)
```
Comment (HarsimranSingh) [inline Gruntfile.js]: Question: when is this used? Has this every worked?

File: Gruntfile.js [hunk 10]
```diff
@@ -455,3 +464,7 @@ module.exports = (grunt) => {
         }
     });
 };
+
+function squash(s) {
```
Comment (HarsimranSingh) [inline Gruntfile.js]: :neckbeard: Maybe a more descriptive name? `stripWhitespace` or something

File: Gruntfile.js [hunk 11]
```diff
@@ -424,14 +431,16 @@ module.exports = (grunt) => {
             // Open-sourced files or new files that are missing a license
             else {
 
+                let squashedFileContent = squash(fileContent);
+
                 // File has commercial license but is not added to the list of commercial files
-                if (fileContent.includes(commercialLic)) {
+                if (squashedFileContent.includes(squash(commercialLic))) {
```
Comment (joneit) [inline Gruntfile.js]: Ignores white space differences.
WebStorm lines up the asterisks whenever I pasted in a multi-line comment.
I realized licenses were never meant to be pasted in manually, so we could do without this.

File: src/shapes.ts [hunk 12]
```diff
@@ -36,13 +36,20 @@ export interface APIMessage {
     payload: any;
 }
 
+export type EverTrue = true;
```
Comment (joneit) [inline src/shapes.ts]: It is semantically sound. I like it. TypeScript will complain:
```js
    ackPayload.success = false;
    nackPayload.success = true;
```

File: src/browser/api_protocol/api_handlers/api_protocol_base.ts (hunk 1 above)
Comment (joneit) [inline src/browser/api_protocol/api_handlers/api_protocol_base.ts]: please explain

File: src/browser/api_protocol/transport_strategy/ack.ts (hunk 6 above)
Comment (joneit) [inline src/browser/api_protocol/transport_strategy/ack.ts]: `error` at least should be, per `NackPayload`'s constructor.
The `APIPayloadNack` (in src/shapes.ts) interface suggests `reason` might be too ...
We are considering merging these interfaces, right?

File: Gruntfile.js (hunk 9 above)
Comment (joneit) [inline Gruntfile.js]: don't know

File: Gruntfile.js (hunk 10 above)
Comment (joneit) [inline Gruntfile.js]: removing this anyway

File: Gruntfile.js (hunk 11 above)
Comment (joneit) [inline Gruntfile.js]: Removing.

File: Gruntfile.js (hunk 9 above)
Comment (rdepena) [inline Gruntfile.js]: why object assign them ? just expand the object then and there.

File: src/browser/api_protocol/api_handlers/clipboard.ts [hunk 13]
```diff
@@ -57,90 +58,52 @@ interface APIMessageClipboardExpanded extends APIMessage {
...
-function clipboardReadText(identity: Identity,
-                           message: APIMessageClipboard,
-                           ack: (payload: APIPayloadAck) => void) {
+function clipboardReadText(identity: Identity, message: APIMessageClipboard, ack: Acker) {
     const {type} = message.payload;
-    ack({
-        success: true,
-        data: clipboard.readText(type)
-    });
+    ack(success(clipboard.readText(type)));
```
Comment (rdepena) [inline src/browser/api_protocol/api_handlers/clipboard.ts]: Revert this file to only include the map changes.

File: src/shapes.ts (hunk 12 above)
Comment (rdepena) [inline src/shapes.ts]: just no.

**Comment by joneit** [comment]
One of the [2 failed tests](https://testing-dashboard.openfin.co/#/app/sessions/api/completed/5924a3105114ac10ca6db455) was _Notification.close (from onShow)_ and the other passed on manual retry.

File: src/shapes.ts [hunk 14]
```diff
@@ -36,13 +36,20 @@ export interface APIMessage {
...
+export type EverFalse = false;
```
Comment (rdepena) [inline src/shapes.ts]: lets not do this.

File: Gruntfile.js (hunk 9 above)
Comment (joneit) [inline Gruntfile.js]: this hash of beautifier options is used elsewhere; I'm mixing it in with the pre-existing option here. (btw, as Sim has pointed out, this grunt sub-task appears to be in disuse.)

File: src/shapes.ts (hunk 14 above)
Comment (joneit) [inline src/shapes.ts]: why is this  a problem? where's the harm? i thought it could only help.

Comment (rdepena) [inline src/shapes.ts]: you need to make a really strong point to redeclare boolean. change it.

**Comment by joneit** [comment]
[Retest](https://testing-dashboard.openfin.co/#/app/sessions/api/completed/59250ded5114ac10ca6db466) had two fails:
* _Notification.close (from onShow)_
* one other that passed on manual retest

**Comment by joneit** [comment]
[Retest](https://testing-dashboard.openfin.co/#/app/sessions/api/completed/592581a25114ac10ca6db467) with 3 fails:
* _Notification.close (from onShow)_
* two others that passed on manual retest

**Comment by joneit** [comment]
I believe I've now made all the requested changes; and retested.

**Comment by datamadic** [comment]
@rdepena can you either OK this one for merge or close? Its OK by me
